Core Platform Upgrades: Ubuntu 20.04, Python 3.8, Node 12, PostgreSQL 12

xitology opened the pull request.

xitology committed:
upgraded to postgresql:11 and latest rexdb/build image

xitology committed:
explicitly install a working version of pip

jayclassless commented on Makefile.src:
```diff
@@ -68,6 +68,7 @@ TEST_PY = ${SRC_PY}
 TOOL_PY = \
 	wheel==0.33.1 \
 	setuptools==41.6.0 \
+	pip==9.0.1 \
```
If we're going to pin pip (which is not a bad idea), why not the most recent instead of one that's 3+ years old?

andreypopp replied:
I remember I had some issues with rex.setup and novel pip versions...

xitology replied:
This will require non-trivial refactoring because newer pip ignores `setup.py` in favor of `pyproject.toml` when the latter is available. The codebase already has a couple of packages with `pyproject.toml`, but ignoring `setup.py` means `rex_init` and `rex_static` are also ignored, which breaks dependency and extension tracing in `rex.core`. That said, we can use moving to `pyproject.toml` as a good opportunity to refactor `rex.core` and get rid of `rex.setup`.

xitology replied:
> I remember I had some issues with rex.setup and novel pip versions...

`pip` deployed by default on ubuntu 19.10 includes a fairly old toml parser, which breaks on a key that starts with a digit.

The latest pip ignores `setup.py` when `pyproject.toml` exists, which breaks dependency and extension tracing in `rex.core`.

jayclassless replied:
Ah, fair enough. I guess we can leave this version for now, but we should try to have this refactor on our near-term to-do list. We could probably bundle this work with reorganizing the package installation process to use a frozen list of packages.

This stuff, plus the [improved dependency resolver in pip](https://pyfound.blogspot.com/2020/03/new-pip-resolver-to-roll-out-this-year.html) coming later this year should make things a bit more stable and/or "normal".

xitology committed:
updated to ubuntu 20.04 with rexdb/build:2020.04.28

xitology committed:
fixed psycopg2 installation failure

jayclassless commented:
Any ideas what's blowing up here?

https://concourse.rexdb.us/builds/5704#L5eaa504c:3120

jayclassless committed:
notebook: Fixed installation of pandas/numpy/matplotlib dependencies

jayclassless commented:
I fixed the installation problems with pandas, etc. Now we're running into an issue with HTSQL on py3.8. Can you look into this @xitology ?

```
        File "/app/src/htsql/src/htsql/__init__.py", line 67, in <module>
          from .core.application import Application as HTSQL
        File "/app/src/htsql/src/htsql/core/__init__.py", line 6, in <module>
          from . import (adapter, addon, application, cache, cmd, connect, context,
        File "/app/src/htsql/src/htsql/core/adapter.py", line 53, in <module>
          class Component(metaclass=ComponentMeta):
        File "/app/src/htsql/src/htsql/core/adapter.py", line 36, in __new__
          code = types.CodeType(code.co_argcount, code.co_kwonlyargcount,
      TypeError: an integer is required (got type bytes)
```

xitology committed:
fixed TypeError in htsql component builder with the latest python

xitology commented:
> I fixed the installation problems with pandas, etc. Now we're running into an issue with HTSQL on py3.8. Can you look into this @xitology ?

Fixed

jayclassless committed:
sms: Upgraded Twilio client library

jayclassless committed:
setup: Fixed tests to work under py3.8

jayclassless committed:
htsql: Fixed use of deprecated cgi.escape() function

jayclassless committed:
action: Fixed usage of deprecated cgi.parse_qs() function

jayclassless committed:
instrument: Fixed syntax warning in a test

jayclassless committed:
graphql: Fixed tests

jayclassless commented:
I've fixed a number of minor py3.8 issues in setup, sms, action, instrument, graphql, and htsql. All tests are passing except for asynctask.

For reasons I haven't been able to nail down yet, when `asynctask-workers` is executed via the `rex.ctl.Ctl()` mechanism we use for testing, the task fails to spawn any of its child processes, so it gets stuck in a loop of trying to restart them. Running `asynctask-workers` normally (outside of `Ctl`) works fine. Help here would be appreciated.

jayclassless committed:
Updated base image version

jayclassless committed:
asynctask: Patched the testing monkeypatch

jayclassless committed:
ci: Make sure we're testing against postgres 11

jayclassless commented:
We're finally green here. Anything else we want to cover in this?

xitology commented:
Since GCP now supports Postgres 12, perhaps we should migrate to it instead of 11. Postgres 12 is also the version distributed with Ubuntu 20.04.

jayclassless commented:
> Since GCP now supports Postgres 12, perhaps we should migrate to it instead of 11. Postgres 12 is also the version distributed with Ubuntu 20.04.

I considered doing that, but when I looked at Azure and AWS, they were only up to 11 (am I wrong?), so I didn't want to trap us.

jayclassless committed:
Forget 11, onward to Postgres 12
